### PR TITLE
Add and expose a `getBookmarkURLForKeyword` API.

### LINKED ISSUE #953: Expose a getBookmarksWithKeyword() API
Firefox for iOS previously had support for bookmarks keyword searches in the AwesomeBar. To continue to support this, we will need the following API:

```
getBookmarksWithKeyword(keyword: String) -> [BookmarkItem]
```

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-153)

### PR DESCRIPTION
**Pull request by tinoism**
Closes #953.

### CODE REVIEW

File: components/places/src/storage/bookmarks.rs [hunk 1]
```diff
@@ -889,6 +889,10 @@ impl<'de> Deserialize<'de> for BookmarkTreeNode {
     }
 }
 
+pub fn bookmarks_get_url_for_keyword(db: &PlacesDb, keyword: &str) -> Result<Option<Url>> {
+    Ok(None)
```
Comment (thomcc) [inline components/places/src/storage/bookmarks.rs]: @linacambridge I'm assuming you just never got to this?

@tinoism If you're interested in finishing this up, this function will need to be completed. (also check with Lina first!)

Comment (tinoism) [reply components/places/src/storage/bookmarks.rs]: Got it!

Comment (linacambridge) [reply components/places/src/storage/bookmarks.rs]: We chatted a bit on Slack! 😄 Thanks for taking this over, @tinoism!

File: .circleci/config.yml [hunk 2]
```diff
@@ -239,7 +239,7 @@ jobs:
       - checkout
       - run:
           name: Lint Bash scripts with shellcheck
-          command: ./automation/lint_bash_scripts.sh
+          command: sh automation/lint_bash_scripts.sh
```
Comment (linacambridge) [inline .circleci/config.yml]: Looks like #1455 is already on master! 🎉

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 3]
```diff
@@ -327,6 +327,18 @@ open class PlacesReaderConnection internal constructor(connHandle: Long) :
         }
     }
 
+    override fun getBookmarkURLForKeyword(keyword: String): String? {
+        val urlPtr = rustCall { error ->
+            LibPlacesFFI.INSTANCE.bookmarks_get_url_for_keyword(this.handle.get(), keyword, error)
+        }
+
+        try {
+            return urlPtr?.getString(0, "utf-8")
```
Comment (linacambridge) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Hmm, this is the second time we're returning null on success. There's a comment at the top of `matchUrl` that suggests breaking this pattern out into a `rustCallForOptString` helper, like `rustCallForString` but with an optional result. How do you feel about adding that to your PR?

File: components/places/src/storage/bookmarks.rs [hunk 4]
```diff
@@ -1408,6 +1425,57 @@ mod tests {
...
+            .into_string();
+
+        conn.execute_named_cached(
+            "INSERT INTO moz_places (guid, url, url_hash) VALUES ('fake_guid___', :url, hash(:url))",
+            &[(":url", &url)],
+        )
+        .expect("should work");
+        let place_id = conn.last_insert_rowid();
+
+        // create a bookmark with keyword 'donut' pointing at it.
```
Comment (linacambridge) [inline components/places/src/storage/bookmarks.rs]: 🍩

File: components/places/android/src/main/java/mozilla/appservices/places/Bookmarks.kt [hunk 5]
```diff
@@ -214,6 +214,17 @@ interface ReadableBookmarksConnection : InterruptibleConnection {
...
+    /**
+     * Returns the URL for the provided search keyword, if one exists.
+     *
+     * @param The search keyword.
+     * @return The bookmarked URL for the keyword, if set.
+     *
+     * @throws OperationInterrupted if this database implements [InterruptibleConnection] and
+     * has its `interrupt()` method called on another thread.
+     */
+    fun getBookmarkURLForKeyword(keyword: String): String?
```
Comment (linacambridge) [inline components/places/android/src/main/java/mozilla/appservices/places/Bookmarks.kt]: Whoops, one more nit: I think the convention we've been following in Kotlin is `Url`, not `URL`. (Swift is fine, looks like we capitalize it as `URL` there).

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt (hunk 3 above)
Comment (tinoism) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Sure, I will add another PR for that issue.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 6]
```diff
@@ -131,6 +131,16 @@ internal inline fun rustCallForString(syncOn: Any, callback: (RustError.ByRefere
     }
 }
 
+@Suppress("TooGenericExceptionThrown")
```
Comment (thomcc) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: You don't throw an exception in this version so you don't need this.

**Comment by linacambridge** [comment]
Thanks @tinoism! :shipit: 🎊

**Pull request merged**